kafka-cron/README.md

```diff
@@ -0,0 +1,161 @@
+
+Read the [Overview of Prometheus](https://prometheus.io/docs/introduction/overview/) if you are not familiar with Prometheus.
+
+The [Prometheus Guide to Instrumenting a Go Application](https://prometheus.io/docs/guides/go-application/) describes how to add metrics to a Golang application.
+
+First, consider:
+ * What kinds of things may go wrong with your system? (it is useful to look at errors your code is handling)
+ * What would users' expectations be of this system?
+ * What metrics can we add that will tell us when the system is not working as intended?
+ * What metrics can we add that might help us to troubleshoot the system and understand how it is operating? Read back through the first three parts of this exercise to try and identify the properties of the system that we might want to know about.
+
+Asking these questions should guide you in designing the metrics that your consumers and producer should export.
+Think about what kinds of problems can happen both in the infrastructure - Kafka, your consumers and producers - and in the submitted jobs. 
+
+Add metrics to your programs. Verify that they work as expected using `curl` or your web browser.
+
+Kafka doesn't export Prometheus metrics natively. However, you can use the official 
+[Prometheus JMX exporter](https://github.com/prometheus/jmx_exporter) to expose its metrics.
+Set this up (it is probably best as another container in your `docker-compose` configuration - you'll need to define a simple Dockerfile for it).
+
+Next, you can add Prometheus, AlertManager, and Grafana, a common monitoring stack, to your `docker-compose` configuration. Here is an example configuration that you can adapt: https://dzlab.github.io/monitoring/2021/12/30/monitoring-stack-docker/. AlertManager is used for notifying operators of unexpected conditions, and Grafana is useful for building dashboards that allow you to troubleshoot and understand your systems operation.
+
+If your system is struggling to run such a complex `docker-compose` system in a performant fashion, you can cut down the number of Kafka topics and consumers that you are running to a minimum (just one producer and consumer/retry consumer pair are fine - don't run sets of these for multiple clusters if your system is under too much load). 
+
+You'll need to set up a Prometheus configuration to scrape your producers and consumers. Prometheus [configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/) is quite complex but you can adapt this [example configuration](https://github.com/prometheus/prometheus/blob/main/documentation/examples/prometheus.yml).
+
+Check that Prometheus is correctly scraping your metrics via http://localhost:9090/metrics.
+
+Next, write an [AlertManager configuration](https://prometheus.io/docs/alerting/latest/alertmanager/) and set up at least one alert. Set up PagerDuty (it has a free trial period available) and get your system to fire an alert, and notify you via PagerDuty.
```

Comment:
I'd maybe drop the PagerDuty part here, and just mention that that's what we'd probably do in real life - I suspect setting it up is a lot more time and possible-things-that-could-go-wrong than it's worth for the learning experience.

Reply:
the other option is to configure email... which is not necessarily easier these days than basic PD setup
i guess we can get them to set up a burner gmail account and do this
https://www.robustperception.io/sending-email-with-the-alertmanager-via-gmail/

Reply:
Sigh, yeah :( What you've got is a good compromise, hopefully!